test/e2e/network/service.go

```diff
@@ -3283,42 +3282,63 @@ var _ = SIGDescribe("ESIPP [Slow]", func() {
 			framework.Failf("Service HealthCheck NodePort still present")
 		}
 
-		endpointNodeMap, err := jig.GetEndpointNodes()
+		config := e2enetwork.NewNetworkingTestConfig(f, false, false)
+
+		epNodes, err := jig.GetEndpointNodeNames()
 		framework.ExpectNoError(err)
-		noEndpointNodeMap := map[string][]string{}
+		// map from name of nodes with endpoint to internal ip
+		endpointNodeMap := make(map[string]string)
+		// map from name of nodes without endpoint to internal ip
+		noEndpointNodeMap := make(map[string]string)
 		for _, n := range nodes.Items {
-			if _, ok := endpointNodeMap[n.Name]; ok {
-				continue
+			ips := e2enode.GetAddresses(&n, v1.NodeInternalIP)
+			if len(ips) == 0 {
+				framework.Failf("No internal ip found for node %s", n.Name)
+			}
+			if epNodes.Has(n.Name) {
+				endpointNodeMap[n.Name] = ips[0]
+			} else {
+				noEndpointNodeMap[n.Name] = ips[0]
 			}
-			noEndpointNodeMap[n.Name] = e2enode.GetAddresses(&n, v1.NodeExternalIP)
 		}
 
 		svcTCPPort := int(svc.Spec.Ports[0].Port)
 		svcNodePort := int(svc.Spec.Ports[0].NodePort)
 		ingressIP := e2eservice.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
 		path := "/clientip"
+		dialCmd := "clientip"
 
 		ginkgo.By(fmt.Sprintf("endpoints present on nodes %v, absent on nodes %v", endpointNodeMap, noEndpointNodeMap))
-		for nodeName, nodeIPs := range noEndpointNodeMap {
-			ginkgo.By(fmt.Sprintf("Checking %v (%v:%v%v) proxies to endpoints on another node", nodeName, nodeIPs[0], svcNodePort, path))
-			GetHTTPContent(nodeIPs[0], svcNodePort, e2eservice.KubeProxyLagTimeout, path)
-		}
-
-		for nodeName, nodeIPs := range endpointNodeMap {
-			ginkgo.By(fmt.Sprintf("checking kube-proxy health check fails on node with endpoint (%s), public IP %s", nodeName, nodeIPs[0]))
-			var body bytes.Buffer
-			pollfn := func() (bool, error) {
-				result := e2enetwork.PokeHTTP(nodeIPs[0], healthCheckNodePort, "/healthz", nil)
-				if result.Code == 0 {
+		for nodeName, nodeIP := range noEndpointNodeMap {
+			ginkgo.By(fmt.Sprintf("Checking %v (%v:%v/%v) proxies to endpoints on another node", nodeName, nodeIP[0], svcNodePort, dialCmd))
+			GetHTTPContentFromTestContainer(config, nodeIP, svcNodePort, e2eservice.KubeProxyLagTimeout, dialCmd)
+		}
+
+		for nodeName, nodeIP := range endpointNodeMap {
```

Comment:
This in total with the previous for may cause timeouts in 5k-node tests, as we're effectively probing every single node in the cluster. Given you touch this already, maybe we should already add a condition that no more than N (=100?) are actually probed?

Reply:
I think it was never the case, we're capping # nodes that we fetch in line 3257:
```
		nodes, err := e2enode.GetBoundedReadySchedulableNodes(cs, e2eservice.MaxNodesForEndpointsTests)
```

But, looking at the code again, I don't think that your code is correct. Neither the old one seems correct.
To be more precise, I think both incorrectly compute the endpointNodeMap. Because of the capping above, the `nodes` will contain a random subset of all nodes. Given that, it's likely (and almost certain in large cluster) that you will get a subset of nodes on which your service is not running. Then the endpointNodeMap will be empty, which won't make the test to fail, but it neither will work correctly. This has to be addressed

Reply:
Nice catch! I want to know if it is guranteed that the cluster always have a node with endpoint and a node without endpoint? If so, I guess we can make the `endpointNodeMap` from the endpoint node names returned by jig, and make the `noEndpointNodeMap` by iterating all nodes until we have selected no more than N nodes without endpoint?

Reply:
Yes, you can assume it's guaranteed. Moreover, we'll have only a single node with the endpoint, so as long as len(nodes) >=2 the code for computing `noEndpointNodeMap` is correct. Only the `endpointNodeMap` computation needs to be fixed and what you proposed makes sens (take it from endpointNodeMap)

Reply:
@mm4tt please check the latest commit

Reply:
The fact that we were never capping number of nodes doesn't mean we shouldn't change that now (I know we were never capping it). But if we want to enable it at 15k-node clusters, we should cap it probably.

Reply:
@wojtek-t if what @mm4tt said was right

> we'll have only a single node with the endpoint

the number of nodes with endpoints is always one, and we get the list of nodes without endpoints by getting bounded nodes first:
```go
nodes, err := e2enode.GetBoundedReadySchedulableNodes(cs, e2eservice.MaxNodesForEndpointsTests /* 3 */)
```
then filter out the node with endpoint:
```go
// it is assumed that there is only a single node with the endpoint
for _, n := range nodes.Items {
    ips := e2enode.GetAddresses(&n, v1.NodeInternalIP)
    if len(ips) < 1 {
        framework.Failf("No internal ip found for node %s", n.Name)
    }
    if _, ok := endpointNodeMap[n.Name]; !ok {
        noEndpointNodeMap[n.Name] = ips[0]
    }
}
```
I guess the number of nodes is already capped?

Reply:
@mm4tt - where do you get it from that there is one node with endpoint? I can't really see any pod and/or endpoint being created in that test (I must be missing something...)

Reply:
Matt pointed me on this offline. So indeed doing that here doesn't make sense.

But it makes sense to do that for "noEndpointMap", because we will have those 5000 iterations there.

Reply:
currently we only iterate over `e2eservice.MaxNodesForEndpointsTests`, which is 3, nodes and filter out the node with endpoint to get `noEndpointMap`. I guess this is OK?

the 5000 iterations actually happens when we want to get a list of nodes with endpoint from node names: https://github.com/kubernetes/kubernetes/pull/91559/files#diff-cd7b1fed55944b797c43381ad94b421fR283-R289

Reply:
perhaps we can using label selector here? like
```go
	namesList := make([]string, 0, nodeNames.Len())
	for name := range nodeNames {
		namesList = append(namesList, name)
	}
	req, err := labels.NewRequirement(v1.LabelHostname, selection.In, namesList)
	s := labels.NewSelector()
	allNodes, err := j.Client.CoreV1().Nodes().List(ctx, metav1.ListOptions{
		LabelSelector: s.Add(*req).String(),
	})
```

Reply:
@wojtek-t I have updated the code as described above, could you check it again?

Reply:
added explicit comment there - with this you risk exceeding allowed length of URL params

It's fine to simply list *all* nodes.

Reply:
That seems fine now.

But I would like to get back to my original comment and ensure that we don't send O(5k) requests in serial here (or rather not here, but for "noEndpoint map" - and instead limit to the first 100 or sth (i.e. min(#such nodes, 100)).

Reply:
according to this line https://github.com/knight42/kubernetes/blob/53f19bc4db12c9071471ab5e30c5cffc983bcf2b/test/e2e/network/service.go#L3257, the number of nodes we iterate at https://github.com/knight42/kubernetes/blob/53f19bc4db12c9071471ab5e30c5cffc983bcf2b/test/e2e/network/service.go#L3301 is `e2eservice.MaxNodesForEndpointsTests`, which is 3.

and I am a bit confused that we don't send any request to apiserver here, `e2enode.GetAddresses()` just extracts the corresponding ip address from the status field of `v1.Node`.

Reply:
Which is fine - because we send request to list nodes before.

Re Bounded nodes - yes, that solves my concern, thanks.